backend/elastic.js

```diff
@@ -113,6 +113,25 @@ class Elastic {
    * @param  {integer} max    The index of last document to retreive
    */
   async search(query, termId, min, max) {
+    // if we know that the query is of the format of a course code, we want to do a very targeted query against subject and classId: otherwise, do a regular query.
+    let courseCodePattern = /^(\s)*\w\w(\w?)(\w?)(\s)*(\d\d\d\d)?(\s)*$/i;
+    let fields = [
+      'class.name^2', // Boost by 2
+      'class.name.autocomplete',
+      'class.subject^4',
+      'class.classId^3',
+      'sections.meetings.profs',
+      'class.crns',
+      'employee.name^2',
+      'employee.emails',
+      'employee.phone',
+    ];
+
+    if (courseCodePattern.test(query)) {
```

Comment:
Need an additional check that the first token in the proposed course code is a subject.

Reply:
Check this out https://github.com/ryanhugh/searchneu/blob/90aa4925a654a928fefe6222515a4148ae82f0e8/backend/search.js#L177
I dumbly deleted it lol.

Reply:
This actually a bit more of a rabbit hole than you may think - the old `checkForSubjectMatch` code depended on `DataLib`, which was also deleted in that same commit. 

DataLib stored pretty much everything in allTerms.json (except allTerms.json wasn't a thing back then) as a variable in the Node.js backend  - including all the subjects. `checkForSubjectMatch` just fetched the list of subjects and then compared to every query.

Reply:
If we wanted to bring this back (which might be better than a regex) I think we could pull the list of subject codes from elasticsearch (or allTerms.json) when the server first starts up.

Reply:
There was something similar done somewhere along the line in one of the really old server.js commits where the backend built up the list of unique subjects while getting subject info to the front end IIRC, my long term memory of this isn't that good though lol

Reply:
Sounds like the right approach. I'm tempted to leave the regex for parsing purposes and just retrieve the subject as the first group and match it against whatever set we have for the available subjects.

Reply:
Oh btw - That link above is the code responsible for taking a search for `cs` and return a list of all the `cs` classes, in order. 

This code here was the code for taking a query like `cs2500` and turning it into a `cs 2500` (with a space). Same file and both of these would run on every query before it was sent off to the old search engine. 

https://github.com/ryanhugh/searchneu/blob/90aa4925a654a928fefe6222515a4148ae82f0e8/backend/search.js#L572-L592